File: api/reports.py
from fastapi import APIRouter, HTTPException, Depends, Query, Response, BackgroundTasks
from typing import List, Optional
from datetime import datetime, timedelta
from models.reports import (
    PerformanceReport,
    CleanupStats,
    ReportSchedule,
    ReportFormat,
    ReportType
)
from services.report_service import ReportService
from services.email_service import EmailService
from auth.auth_service import get_current_user
import logging
import pandas as pd
import io

router = APIRouter()
logger = logging.getLogger(__name__)
report_service = ReportService()
email_service = EmailService()

@router.get("/performance", response_model=PerformanceReport)
async def get_performance_report(
    start_date: datetime = Query(...),
    end_date: datetime = Query(...),
    format: ReportFormat = ReportFormat.JSON,
    current_user = Depends(get_current_user)
):
    try:
        report_data = await report_service.generate_performance_report(
            start_date,
            end_date
        )

        if format == ReportFormat.EXCEL:
            excel_data = await report_service.generate_excel_report(report_data)
            return Response(
                content=excel_data,
                media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                headers={"Content-Disposition": "attachment; filename=performance-report.xlsx"}
            )
        elif format == ReportFormat.PDF:
            pdf_data = await report_service.generate_pdf_report(report_data)
            return Response(
                content=pdf_data,
                media_type="application/pdf",
                headers={"Content-Disposition": "attachment; filename=performance-report.pdf"}
            )

        return report_data
    except Exception as e:
        logger.error(f"Failed to generate performance report: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to generate report")

@router.get("/cleanup-stats", response_model=CleanupStats)
async def get_cleanup_statistics(
    period: str = "30d",
    current_user = Depends(get_current_user)
):
    try:
        return await report_service.get_cleanup_stats(period)
    except Exception as e:
        logger.error(f"Failed to get cleanup stats: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to get cleanup statistics")

@router.post("/schedule")
async def schedule_report(
    report_schedule: ReportSchedule,
    background_tasks: BackgroundTasks,
    current_user = Depends(get_current_user)
):
    try:
        job_id = await report_service.schedule_report(
            report_schedule,
            background_tasks
        )
        return {"message": "Report scheduled successfully", "job_id": job_id}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to schedule report: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to schedule report") 

File: api/reports.js
const express = require('express');
const router = express.Router();
const { requirePermission } = require('../auth/roleMiddleware');
const { PERMISSIONS } = require('../config/roles');
const logger = require('../utils/logger');
const ExcelJS = require('exceljs');
const PDFDocument = require('pdfkit');

// Get performance report
router.get('/performance', 
  requirePermission(PERMISSIONS.VIEW_LOGS),
  async (req, res) => {
    try {
      const { startDate, endDate, format = 'json' } = req.query;
      const report = await generatePerformanceReport(startDate, endDate);

      if (format === 'excel') {
        const buffer = await generateExcelReport(report);
        res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        res.setHeader('Content-Disposition', 'attachment; filename=performance-report.xlsx');
        return res.send(buffer);
      } else if (format === 'pdf') {
        res.setHeader('Content-Type', 'application/pdf');
        res.setHeader('Content-Disposition', 'attachment; filename=performance-report.pdf');
        return generatePDFReport(report, res);
      }

      res.json(report);
    } catch (error) {
      logger.error('Failed to generate performance report:', error);
      res.status(500).json({ error: 'Failed to generate report' });
    }
});

// Get cleanup statistics
router.get('/cleanup-stats',
  requirePermission(PERMISSIONS.VIEW_LOGS),
  async (req, res) => {
    try {
      const { period = '30d' } = req.query;
      const stats = await getCleanupStats(period);
      res.json(stats);
    } catch (error) {
      logger.error('Failed to get cleanup stats:', error);
      res.status(500).json({ error: 'Failed to get cleanup statistics' });
    }
});

// Schedule report generation
router.post('/schedule',
  requirePermission(PERMISSIONS.SYSTEM_CONFIG),
  async (req, res) => {
    try {
      const { 
        reportType,
        schedule,
        recipients,
        format = 'pdf'
      } = req.body;

      const job = await scheduleReport({
        reportType,
        schedule,
        recipients,
        format
      });

      res.json({ 
        message: 'Report scheduled successfully',
        jobId: job.id
      });
    } catch (error) {
      logger.error('Failed to schedule report:', error);
      res.status(400).json({ error: error.message });
    }
});

async function generatePerformanceReport(startDate, endDate) {
  const metrics = await getPerformanceMetrics(startDate, endDate);
  const cleanupStats = await getCleanupStats(startDate, endDate);
  const diskUsage = await getDiskUsageHistory(startDate, endDate);

  return {
    period: {
      start: startDate,
      end: endDate
    },
    summary: {
      totalCleanups: cleanupStats.totalJobs,
      successRate: cleanupStats.successRate,
      totalRecordsArchived: cleanupStats.totalRecords,
      averageDuration: cleanupStats.averageDuration
    },
    performance: {
      averageCPU: metrics.averageCPU,
      peakCPU: metrics.peakCPU,
      averageMemory: metrics.averageMemory,
      peakMemory: metrics.peakMemory
    },
    storage: {
      averageDiskUsage: diskUsage.average,
      peakDiskUsage: diskUsage.peak,
      spaceReclaimed: diskUsage.reclaimed
    },
    trends: {
      dailyStats: metrics.dailyStats,
      weeklyStats: metrics.weeklyStats
    }
  };
}

async function generateExcelReport(data) {
  const workbook = new ExcelJS.Workbook();
  const worksheet = workbook.addWorksheet('Performance Report');

  // Add summary section
  worksheet.addRow(['Performance Report Summary']);
  worksheet.addRow(['Period', `${data.period.start} to ${data.period.end}`]);
  worksheet.addRow(['Total Cleanups', data.summary.totalCleanups]);
  worksheet.addRow(['Success Rate', `${data.summary.successRate}%`]);
  worksheet.addRow(['Records Archived', data.summary.totalRecordsArchived]);

  // Add performance metrics
  worksheet.addRow([]);
  worksheet.addRow(['Performance Metrics']);
  worksheet.addRow(['Average CPU', `${data.performance.averageCPU}%`]);
  worksheet.addRow(['Peak CPU', `${data.performance.peakCPU}%`]);
  worksheet.addRow(['Average Memory', `${data.performance.averageMemory}MB`]);
  worksheet.addRow(['Peak Memory', `${data.performance.peakMemory}MB`]);

  // Add daily stats
  worksheet.addRow([]);
  worksheet.addRow(['Daily Statistics']);
  worksheet.addRow(['Date', 'Jobs Run', 'Success Rate', 'Records Processed']);
  data.trends.dailyStats.forEach(stat => {
    worksheet.addRow([
      stat.date,
      stat.jobsRun,
      `${stat.successRate}%`,
      stat.recordsProcessed
    ]);
  });

  return await workbook.xlsx.writeBuffer();
}

async function generatePDFReport(data, res) {
  const doc = new PDFDocument();
  doc.pipe(res);

  // Add title
  doc.fontSize(20).text('Performance Report', { align: 'center' });
  doc.moveDown();

  // Add summary section
  doc.fontSize(16).text('Summary');
  doc.fontSize(12)
    .text(`Period: ${data.period.start} to ${data.period.end}`)
    .text(`Total Cleanups: ${data.summary.totalCleanups}`)
    .text(`Success Rate: ${data.summary.successRate}%`)
    .text(`Records Archived: ${data.summary.totalRecordsArchived}`);
  doc.moveDown();

  // Add performance section
  doc.fontSize(16).text('Performance Metrics');
  doc.fontSize(12)
    .text(`Average CPU: ${data.performance.averageCPU}%`)
    .text(`Peak CPU: ${data.performance.peakCPU}%`)
    .text(`Average Memory: ${data.performance.averageMemory}MB`)
    .text(`Peak Memory: ${data.performance.peakMemory}MB`);

  // Add charts and graphs
  // ... Add visualization logic here

  doc.end();
}

module.exports = router; 